### PR TITLE
Bolt agent added to hello metadata

### PR DESCRIPTION
**Pull request by ConorNeo**

### CODE REVIEW

File: packages/bolt-connection/src/bolt/bolt-protocol-v3.js [hunk 1]
```diff
@@ -78,7 +78,7 @@ export default class BoltProtocol extends BoltProtocolV2 {
     // passing notification filter on this protocol version throws an error
     assertNotificationFilterIsEmpty(notificationFilter, this._onProtocolError, observer)
 
-    this.write(RequestMessage.hello(userAgent, authToken), observer, true)
+    this.write(RequestMessage.hello(userAgent === '' || userAgent == null ? boltAgent : userAgent, authToken), observer, true)
```
Comment (bigmontz) [inline packages/bolt-connection/src/bolt/bolt-protocol-v3.js]: Should we treat `''` as null?

File: packages/bolt-connection/src/bolt/request-message.js [hunk 2]
```diff
@@ -191,6 +191,38 @@ export default class RequestMessage {
...
+  /**
+   * Create a new HELLO message.
+   * @param {string} userAgent the user agent.
+   * @param {string} boltAgent the bolt agent.
+   * @param {NotificationFilter} notificationFilter the notification filter configured
+   * @param {Object} routing server side routing, set to routing context to turn on server side routing (> 4.1)
+   * @return {RequestMessage} new HELLO message.
+   */
+  static hello5x3 (userAgent, boltAgent, notificationFilter = null, routing = null) {
+    const metadata = { user_agent: userAgent, bolt_agent: boltAgent }
```
Comment (bigmontz) [inline packages/bolt-connection/src/bolt/request-message.js]: maybe we should omit `user_agent` if it's `null` or `undefined`.

File: packages/core/src/internal/bolt-agent/browser/BoltAgent.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+/**
```
Comment (bigmontz) [inline packages/core/src/internal/bolt-agent/browser/BoltAgent.ts]: the file names are created lowercase and separated by `-`. So, this file should be called `bolt-agent.ts` (or move this implementation to be directly in the `index.ts`)

File: packages/neo4j-driver-deno/generate.ts [hunk 4]
```diff
@@ -129,15 +129,15 @@ async function copyAndTransform(inDir: string, outDir: string) {
...
       // Replace the "node channel" with the "deno channel", since Deno supports different APIs
-      if (inPath.endsWith("channel/index.js")) {
+      // Replace the "node boltAgent" with the "deno boltAgent", since Deno supports different APIs
+      if (inPath.endsWith("channel/index.js") || inPath.endsWith("bolt-agent/index.ts")) {
         contents = contents.replace(
-          `export * from './node/index.js'`,
-          `export * from './deno/index.js'`,
+          `export * from './node/index'`,
+          `export * from './deno/index'`,
         );
```
Comment (bigmontz) [inline packages/neo4j-driver-deno/generate.ts]: Did you test the deno driver in testkit? I don't know if the `export` replace will work without the extension.

File: packages/neo4j-driver-lite/src/index.ts [hunk 5]
```diff
@@ -376,7 +379,7 @@ async function hasReachableServer (url: string, config?: Pick<Config, 'logging'>
   }
 }
 
-const USER_AGENT: string = 'neo4j-javascript/' + VERSION
+const BOLT_AGENT = internal.boltAgent.BoltAgent.fromVersion('neo4j-javascript/' + VERSION)
```
Comment (bigmontz) [inline packages/neo4j-driver-lite/src/index.ts]: Maybe move the constant definition before it usage. And I think this line might be `internal.boltAgent.fromVersion`

File: packages/neo4j-driver/src/index.js [hunk 6]
```diff
@@ -342,7 +343,7 @@ async function hasReachableServer (url, config) {
   }
 }
 
-const USER_AGENT = 'neo4j-javascript/' + VERSION
+const BOLT_AGENT = internal.boltAgent.fromVersion('neo4j-javascript/' + VERSION)
```
Comment (bigmontz) [inline packages/neo4j-driver/src/index.js]: Maybe move the definition to done before the usage (i know this was already this way, but I think it worth the small refactory).

**Review by bigmontz — COMMENTED** [review]
This is looking good so far. 

I think it might worth added tests for old protocols to check if they are using the `boltAgent` as fallback for the case the user is not setting the `userAgent`.

File: packages/bolt-connection/src/bolt/bolt-protocol-v3.js (hunk 1 above)
Comment (ConorNeo) [reply packages/bolt-connection/src/bolt/bolt-protocol-v3.js]: if i dont it causes a testkit test to fail

File: packages/bolt-connection/src/bolt/request-message.js (hunk 2 above)
Comment (ConorNeo) [reply packages/bolt-connection/src/bolt/request-message.js]: yeah makes sense!

**Review by bigmontz — APPROVED** [review]
🥇

**Pull request merged**